### PR TITLE
OPENEUROPA-2873: Style Facts and figures paragraph.

### PR DESCRIPTION
**Pull request by 22Alexandra**
### Description

Create new Facts and figures pattern based on ECL component and use it to style the Facts and figures paragraph.

### Change log

- Added: facts and figures pattern
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: oe_theme.theme [hunk 1]
```diff
@@ -1315,3 +1315,106 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
 }
+
+/**
+ * Implements hook_preprocess_paragraph().
+ */
+function oe_theme_preprocess_paragraph__oe_facts_figures(array &$variables): void {
+  $paragraph = $variables['paragraph'];
+  $variables['title'] = $paragraph->get('field_oe_title')->value;
+  $variables['view_all'][0]['path'] = $paragraph->get('field_oe_link')->first()->getUrl();
+  $variables['view_all'][0]['label'] = $paragraph->get('field_oe_link')->first()->get('title')->getValue();
```
Comment (ademarco) [inline oe_theme.theme]: You don't need to have an extra array in a preview, you can simply do:

```
  $variables['view_all']['path'] = $paragraph->get('field_oe_link')->first()->getUrl();
  $variables['view_all']['label'] = $paragraph->get('field_oe_link')->first()->get('title')->getValue();
```

File: oe_theme.theme [hunk 2]
```diff
@@ -1315,3 +1315,106 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
+/**
+ * Helper function to convert icon names.
+ *
+ * @param string $icon_name
+ *   The icon name to be converted.
+ *
+ * @return string
+ *   The converted icon name or digital if the icon name is not supported.
+ */
+function _oe_theme_icons_convert(string $icon_name): string {
```
Comment (ademarco) [inline oe_theme.theme]: On a second thought let's turn this into a Twig filter, we did the same for file icon. How to do:

In \Drupal\oe_theme_helper\TwigExtension\Filters add:

```php
  public function getFilters(): array {
    return [
      ....
      new \Twig_SimpleFilter('to_icon', [$this, 'toIcon']),
    ];
  }
```

Then convert this function into a method of the `Filters` class and use it directly in the Twig template with `|to_icon`.

File: oe_theme.theme [hunk 3]
```diff
@@ -1315,3 +1315,106 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
+  $paragraph = $variables['paragraph'];
+  $variables['title'] = $paragraph->get('field_oe_title')->value;
+  $variables['view_all'][0]['path'] = $paragraph->get('field_oe_link')->first()->getUrl();
+  $variables['view_all'][0]['label'] = $paragraph->get('field_oe_link')->first()->get('title')->getValue();
+  $variables['items'] = [];
+
+  foreach ($paragraph->get('field_oe_paragraphs')->referencedEntities() as $sub_paragraph) {
+    /** @var \Drupal\paragraphs\Entity\Paragraph $sub_paragraph */
+    $variables['items'][] = [
+      'icon' => _oe_theme_icons_convert($sub_paragraph->get('field_oe_icon')->value),
```
Comment (ademarco) [inline oe_theme.theme]: See below, pass the raw value here and use the new Twig filter to convert icons into ECL icon names.

File: templates/paragraphs/paragraph--oe-facts-figures.html.twig [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * @file
+ * Theme override to display the 'Facts and figures' #type paragraph.
+ *
+ * @see ./modules/contrib/paragraphs/templates/paragraph.html.twig
+ */
+#}
+<h3 class="ecl-u-type-heading-3">
+  {{ title }}
+</h3>
```
Comment (ademarco) [inline templates/paragraphs/paragraph--oe-facts-figures.html.twig]: Only print this if title is set.

File: templates/patterns/facts_figures/facts_figures.ui_patterns.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+          value: "00.0 million"
+          title: "Nam lacinia nisl eget diam mattis"
+          description: "Sed efficitur bibendum rutrum. Nunc feugiat congue augue ac consectetur."
+    view_all:
+      type: "array"
+      label: "View all"
+      description: "View all link"
+      preview:
+        - label: "View all metrics"
+          path: "/example"
```
Comment (ademarco) [inline templates/patterns/facts_figures/facts_figures.ui_patterns.yml]: Just use here:
```yml
      preview:
          label: "View all metrics"
          path: "/example"
```
So you can get rid of the extra array.

File: templates/patterns/facts_figures/pattern-facts-figures.html.twig [hunk 6]
```diff
@@ -0,0 +1,26 @@
+{#
+/**
+ * @file
+ * Facts and figures component.
+ */
+#}
+{% set _link = {'path' : view_all.0.path, 'label' : view_all.0.label} %}
```
Comment (ademarco) [inline templates/patterns/facts_figures/pattern-facts-figures.html.twig]: Following comment above this becomes:
```twig
 {% set _link = {'path' : view_all.path, 'label' : view_all.label} %}
```

File: templates/patterns/facts_figures/pattern-facts-figures.html.twig [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+{% set _items = [] %}
+{% for _item in items %}
+  {% set _items = _items|merge([{
+    icon: {
+      path: ecl_icon_path,
+      name: _item.icon,
```
Comment (ademarco) [inline templates/patterns/facts_figures/pattern-facts-figures.html.twig]: Use here ```       name: _item.icon|to_icon,```

File: templates/patterns/facts_figures/facts_figures.ui_patterns.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+          title: "Aenean dapibus"
+          description: "Aliquam lacinia diam eu sem malesuada, in interdum ante bibendum."
+        - icon: "notifications--success"
+          value: "00.0 million"
+          title: "Aliquam faucibus nulla eget eleifend"
+          description: "Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Duis nec lectus tortor."
+        - icon: "ui--external"
+          value: "00.0 million"
+          title: "Nam lacinia nisl eget diam mattis"
+          description: "Sed efficitur bibendum rutrum. Nunc feugiat congue augue ac consectetur."
```
Comment (ademarco) [inline templates/patterns/facts_figures/facts_figures.ui_patterns.yml:33]: We shouldn't use ECL names here, but our names, as ECL might change them one day. If that happens then we will change them in the `to_icon` filter only.

Let's use simple names here, like those we used on the paragraph, so something like:

```yml
      preview:
        - icon: "instagram"
          value: "00.0 million"
          title: "Lorem ipsum"
          description: "Nunc condimentum sapien ut nibh finibus suscipit vitae at justo. Morbi quis odio faucibus, commodo tortor id, elementum libero."
        - icon: "twitter"
          value: "00.0 million"
          title: "Sed hendrerit"
          description: "Turpis varius congue venenatis, erat dui feugiat felis."
```

**Review by ademarco — CHANGES_REQUESTED** [review]

File: modules/oe_theme_helper/src/TwigExtension/Filters.php [hunk 9]
```diff
@@ -219,4 +220,86 @@ public function toEclAttributes($attributes): array {
...
+      'rounded-arrow' => 'ui--rounded-arrow',
+      'solid-arrow' => 'ui--solid-arrow',
+    ];
+
+    if (array_key_exists($icon, $icons)) {
+      return $icons[$icon];
+    }
+    else {
+      return 'general--digital';
+    }
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/Filters.php]: We can simply use:
```php
return $icons[$icon] ?? 'general--digital';
```

File: oe_theme.theme [hunk 10]
```diff
@@ -1315,3 +1315,24 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
 }
+
+/**
+ * Implements hook_preprocess_paragraph().
+ */
+function oe_theme_preprocess_paragraph__oe_facts_figures(array &$variables): void {
+  $paragraph = $variables['paragraph'];
+  $variables['title'] = $paragraph->get('field_oe_title')->value;
+  $variables['view_all']['path'] = $paragraph->get('field_oe_link')->first()->getUrl();
+  $variables['view_all']['label'] = $paragraph->get('field_oe_link')->first()->get('title')->getValue();
```
Comment (ademarco) [inline oe_theme.theme]: As link field is [not required](https://github.com/openeuropa/oe_paragraphs/pull/85/files#diff-2323e87dad8419a34eb506df52b554f1R15) we need to check if that's not empty first.

File: oe_theme.theme [hunk 11]
```diff
@@ -1315,3 +1315,24 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
+  $variables['items'] = [];
+
+  foreach ($paragraph->get('field_oe_paragraphs')->referencedEntities() as $sub_paragraph) {
+    /** @var \Drupal\paragraphs\Entity\Paragraph $sub_paragraph */
```
Comment (ademarco) [inline oe_theme.theme]: We need to get translation from context here, like we did on `oe_theme_preprocess_paragraph__oe_content_row__variant_inpage_navigation()`:
```php
    // Get sub-paragraph translation.
    $sub_paragraph = \Drupal::service('entity.repository')
      ->getTranslationFromContext($sub_paragraph, $paragraph->language()->getId());
```

File: tests/Kernel/Paragraphs/ParagraphsTest.php [hunk 12]
```diff
@@ -561,4 +561,66 @@ public function testContextualNavigation(): void {
...
+    $this->assertEquals('Fact description 1', trim($crawler->filter('div.ecl-fact-figures__item:nth-child(1) div.ecl-fact-figures__description')->text()));
+    $this->assertEquals('Fact description 2', trim($crawler->filter('div.ecl-fact-figures__item:nth-child(2) div.ecl-fact-figures__description')->text()));
+    $this->assertEquals('Fact description 3', trim($crawler->filter('div.ecl-fact-figures__item:nth-child(3) div.ecl-fact-figures__description')->text()));
+
+    $link = $crawler->filter('div.ecl-fact-figures__view-all a.ecl-link.ecl-link--standalone.ecl-fact-figures__view-all-link');
+    $actual = $link->text();
+    $this->assertEquals('View all metrics', trim($actual));
+    $actual = $link->attr('href');
+    $this->assertEquals('http://www.example.com/', trim($actual));
+  }
```
Comment (ademarco) [inline tests/Kernel/Paragraphs/ParagraphsTest.php:636]: We need to test that, when view all is empty, then we don't print any button.

File: templates/patterns/facts_figures/pattern-facts-figures.html.twig [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    title: _item.title,
+    description: _item.description,
+  }]) %}
+{% endfor %}
+
+{% include '@ecl-twig/fact-figures' with {
+  'items': _items,
+  "view_all":{
+    "link": _link
+  }
```
Comment (ademarco) [inline templates/patterns/facts_figures/pattern-facts-figures.html.twig:24]: We should print view all only if not empty.

File: templates/patterns/facts_figures/pattern-facts-figures.html.twig [hunk 14]
```diff
@@ -0,0 +1,26 @@
+{#
+/**
+ * @file
+ * Facts and figures component.
```
Comment (ademarco) [inline templates/patterns/facts_figures/pattern-facts-figures.html.twig]:
```
  * Facts and figures pattern template.
``

File: modules/oe_theme_helper/src/TwigExtension/Filters.php [hunk 15]
```diff
@@ -219,4 +220,86 @@ public function toEclAttributes($attributes): array {
     return $ecl_attributes;
   }
 
+  /**
+   * Convert icon names to the ECL supported names.
+   *
+   * @param string $icon
+   *   The icon to be to be converted.
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/Filters.php]: Change into:
```
The icon to be converted.
```

File: modules/oe_theme_helper/src/TwigExtension/Filters.php [hunk 16]
```diff
@@ -219,4 +220,86 @@ public function toEclAttributes($attributes): array {
...
+   *
+   * @return string
+   *   The converted icon name or digital if the icon name is not supported.
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/Filters.php]:
```
    *   The converted icon name or digital if the icon name, defaults to "general--digital".
```

**Review by ademarco — CHANGES_REQUESTED** [review]

File: oe_theme.theme [hunk 17]
```diff
@@ -1321,13 +1321,20 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
  */
 function oe_theme_preprocess_paragraph__oe_facts_figures(array &$variables): void {
   $paragraph = $variables['paragraph'];
-  $variables['title'] = $paragraph->get('field_oe_title')->value;
-  $variables['view_all']['path'] = $paragraph->get('field_oe_link')->first()->getUrl();
-  $variables['view_all']['label'] = $paragraph->get('field_oe_link')->first()->get('title')->getValue();
+  if (!empty($paragraph->get('field_oe_title')->value)) {
```
Comment (ademarco) [inline oe_theme.theme]: no need to check here as the field is required.

File: oe_theme.theme [hunk 18]
```diff
@@ -1321,13 +1321,20 @@ function _oe_theme_trim_textfield($text, int $size = 150) {
...
+    $variables['title'] = $paragraph->get('field_oe_title')->value;
+  }
+  if (!empty($paragraph->get('field_oe_link')->first())) {
```
Comment (ademarco) [inline oe_theme.theme]: Use `->isEmpty()` method:
```php
   if (!$paragraph->get('field_oe_title')->isEmpty()) {
```

**Review by ademarco — CHANGES_REQUESTED** [review]

**Review by ademarco — CHANGES_REQUESTED** [review]
I get default icon also when selecting a specific one, like in this case:

image


image

I get the same icon when I select "Arrow down". I guess something is wrong with mapping.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+use Drupal\Core\Template\Attribute;
+use Drupal\oe_theme_helper\EuropeanUnionLanguages;
+
+/**
+ * Collection of extra Twig filters.
+ *
+ * We don't enforce any strict type checking on filters' arguments as they are
+ * coming straight from Twig templates.
+ */
+class TwigExtension extends \Twig_Extension {
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:18]: Update description after renaming.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFunctions(): array {
+    return parent::getFunctions() + [
+      new \Twig_SimpleFunction('to_ecl_icon', [$this, 'toEclIcon']),
+    ];
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php:58]: Let's only use
```php
    return [
      new \Twig_SimpleFunction('to_ecl_icon', [$this, 'toEclIcon']),
    ];
```

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+        'transform' => 'rotate-90',
+      ],
+    ];
+
+    if (array_key_exists($icon, $icons)) {
+      $icons[$icon]['path'] = $path;
+      return $icons[$icon];
+    }
+
+    return 'general--digital';
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: This is wrong right? it needs to always return an ECL icon array.

File: modules/oe_theme_helper/src/TwigExtension/TwigExtension.php [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+   *   The icon to be converted.
+   *
+   * @return array
+   *   Icon array for ECL components containing icon name, path and rotation.
+   */
+  public function toEclIcon(string $icon): array {
+    // ECL Icons path.
+    $path = base_path() . drupal_get_path('theme', 'oe_theme') .
+      '/dist/' . theme_get_setting('component_library') .
+      '/images/icons/sprites/icons.svg';
```
Comment (ademarco) [inline modules/oe_theme_helper/src/TwigExtension/TwigExtension.php]: We can request Twig context when declaring the function, after that we can simply access the variable set in `oe_theme_preprocess()` via the context, so that we don't have to duplicate this code here.

```php
class TwigFunction
{
    protected $name;
    protected $callable;
    protected $options;
    protected $arguments = [];

    public function __construct($name, $callable, array $options = [])
    {
        $this->name = $name;
        $this->callable = $callable;
        $this->options = array_merge([
            'needs_environment' => false,
            'needs_context' => false,
            'is_variadic' => false,
            'is_safe' => null,
            'is_safe_callback' => null,
            'node_class' => '\Twig\Node\Expression\FunctionExpression',
            'deprecated' => false,
            'alternative' => null,
        ], $options);
    }
```

**Review by ademarco — CHANGES_REQUESTED** [review]

**Review by ademarco — DISMISSED** [review]

File: oe_theme.theme (hunk 10 above)
Comment (sergepavle) [reply oe_theme.theme]: Yes. still missed $paragraph->get('field_oe_link')->isEmpty()

File: templates/patterns/facts_figures/pattern-facts-figures.html.twig (hunk 13 above)
Comment (sergepavle) [reply templates/patterns/facts_figures/pattern-facts-figures.html.twig:24]: Still missed.

File: oe_theme.theme (hunk 17 above)
Comment (sergepavle) [reply oe_theme.theme]: We are still checking if this fill is empty.

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php [hunk 23]
```diff
@@ -7,17 +7,17 @@
 use Drupal\Core\Language\LanguageInterface;
 use Drupal\Core\Language\LanguageManagerInterface;
 use Drupal\Core\Template\Loader\StringLoader;
-use Drupal\oe_theme_helper\TwigExtension\Filters;
+use Drupal\oe_theme_helper\TwigExtension\TwigExtension;
 use Drupal\Tests\UnitTestCase;
 
 /**
  * Tests for the custom Twig filters extension.
```
Comment (sergepavle) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php]: needs to update

File: modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php [hunk 24]
```diff
@@ -7,17 +7,17 @@
...
  *
  * @group oe_theme_helper
  *
- * @coversDefaultClass \Drupal\oe_theme_helper\TwigExtension\Filters
+ * @coversDefaultClass \Drupal\oe_theme_helper\TwigExtension\TwigExtension
  */
-class TwigFiltersTest extends UnitTestCase {
+class TwigExtensionTest extends UnitTestCase {
```
Comment (sergepavle) [inline modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php:20]: As I can see here missed test coverage for new toEclIcon() method.
@ademarco do we need to do this within this PR?

Comment (ademarco) [reply modules/oe_theme_helper/tests/src/Unit/TwigExtensionTest.php:20]: Yeah maybe let's just add a very simple one that tests that non existing icon will get the default result.

**Review by sergepavle — APPROVED** [review]

**Review by imanoleguskiza — APPROVED** [review]
